docs: add DR on the technical direction of decentralized identity for the EDC (#96)
eclipse-edc/docs: docs/developer/decision-records/2023-09-06-identity-trust/README.md

# Technical Direction for a Robust Decentralized EDC Identity and Trust System

## Decision

This decision record establishes the technical direction for developing an EDC identity and trust system that can be
used in
place of the existing OAuth2-based system when required. This system will be built on existing EDC infrastructure,
including the `Identity Hub.`

## Rationale

Identity and trust are the cornerstones of a dataspace. Currently, the EDC has production-quality support for the
exchange of identity and claims information using
the [OAuth 2 Client Credentials Grant](https://datatracker.ietf.org/doc/html/rfc6749#section-4.4). For many use cases,
relying on a centralized OAuth 2 Authorization Server is sufficient. However, there are a class of uses-cases which
require a decentralized approach to exchanging identity and claims data. For example, operational continuity
requirements may determine that a centralized Authentication Server poses too great a risk of service disruption. Or,

> privacy rules may dictate that claims verification must be done without requiring both the verifier and verified being

Reviewer: ```suggestion
privacy rules may dictate that claims verification must be done without requiring both the verifying and verified party being
```

known to a third-party identity provider.

The decentralized identity and trust system will enable dataspace participants to securely communicate and validate
credentials using
the [Dataspace Protocol Specifications](https://github.com/International-Data-Spaces-Association/ids-specification) (
DSP) in a way that preserves privacy and limits the possibility of network disruption.

## Approach

> The EDC identity and trust system will define a set of message interactions (termed "protocols") for the following:

Reviewer 2: "The EDC identity and trust system will define a set of message interactions" together with "For convenience" (line 37) could be interpreted as "the EDC are defining the specifications and just locate it in another repo"... This DR should highlight that the EDC are the place for the implementation and the EDWG (or currently Tractus) for specification.
```suggestion
The EDC identity and trust system will implement a set of message interactions (termed "protocols") for the following:
```

- **Self-Issued Identity Tokens**: Obtaining and cryptographically validating self-issued tokens that contain identity
  claims
- **Verifiable Credential Presentation**: Communicating Verifiable Credentials (VC), Verifiable Presentations (VP), and
  the cryptographic material needed to validate them
- **Verifiable Credential Issuance**: Issuing and renewing Verifiable Credentials and providing a mechanism for
  communicating credential revocation

> For convenience (and because not all stakeholders are EDC committers), the protocol specifications will be temporarily

Reviewer 2: ```suggestion
The protocol specifications are currently
```

> maintained in the [Eclipse Tractus-X SSI repository](https://github.com/eclipse-tractusx/ssi-docu). These specifications

Reviewer 2: ```suggestion
maintained in the [Eclipse Tractus-X SSI repository](https://github.com/eclipse-tractusx/ssi-docu), and
```

> will be transferred to the Eclipse Dataspace Working Group as

Reviewer 2: ```suggestion
are going to be transferred to the Eclipse Dataspace Working Group as
```

described [here](#protocol-specifications-and-the-eclipse-dataspace-working-group).

### Self-Issued Identity Tokens

The shape of self-issued identity tokens and protocol for obtaining them will be based on
the [Self-Issued OpenID Provider v2 specification](https://openid.net/specs/openid-connect-self-issued-v2-1_0.html#section-1.1) (
SIOPv2).

### Verifiable Credential Presentation

> Data providers will often require a set of Verifiable Presentations (i.e. the presentation of a Verifiable Credential)

Reviewer 3: That sentence looks a bit weird. 
The data provider will determine whether a given consumer can access its asset based on its Verifiable Presentation.

Author: It reads: "Data providers will often require a set of Verifiable Presentations (i.e. the presentation of a Verifiable Credential) to access an asset." Does that sound weird because to me it is a direct statement?

Reviewer 3: Data provider does not access an asset. 
Data provider grants (or not) access to an asset based on the consumer VP

### Verifiable Credential Issuance

Dataspaces will require an interoperable way for participant agent systems to request credential issuance, renew
credentials, and check for credential revocation. The Verifiable Credential Issuance will define a RESTful API over
HTTPS that uses the [W3C Decentralized Identifiers specification](https://www.w3.org/TR/did-core/) and
the [W3C Status List specification](https://www.w3.org/TR/vc-status-list/).

Note that
the [OpenID for Verifiable Credential Issuance specification](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-introduction)
was considered, but determined not to be sufficient, due to its focus on end-user interactions and lack of support for
long-running issuance processes that may involve human interaction.

### Protocol Specifications and The Eclipse Dataspace Working Group

The protocol specifications (i.e. documents that define the message exchanges) for the Identity and Trust System will be
transferred to the [Eclipse Dataspace Working Group](https://www.eclipse.org/org/workinggroups/dataspace-charter.php)

> for formal standardization. The EDC will track changes to the protocol specifications as they undergo standardization.

Reviewer 4: Well, it is not a formal standardization, but preparatory work under a well-accepted rules in terms of transparency, governance, and collaboration to contribute it to a formal standarization body after reaching an adequate maturity.

Author: OK I've removed formal.

## Impact

The decentralized Identity and Trust System will have negligible architectural impact, as this work will be an extension
of existing EDC services. The most notable impact will be on the wire protocol used by the `Identity Hub.`

### Identity Hub

The `Identity Hub` is currently based on an early version of
the [Decentralized Web Nodes specification](https://identity.foundation/decentralized-web-node/spec/) (DWN).
The `Identity Hub` lacks several important features including VC revocation, key rotation, and permission-based access.
Although the DWN standard specifies a permission-based access method, it is not compatible with OAuth 2 and SIOPv2.
Moreover, the goal of the DWN specification is to define a generic credential storage and message relay system. As such,
it does not address key topics including credential metadata exchange. The DWN specification will therefore be replaced

> by the above Eclipse protocol specifications, which will be aligned with standards including SIOPv2, DIF Presentation

Reviewer: ```suggestion
by the protocol specifications mentioned above, which will be aligned with standards including SIOPv2, DIF Presentation
```
